mtx/mtx.1
.\" mtx.1  Document copyright 2000 Eric Lee Green
.\"  Program Copyright 1996, 1997 Leonard Zubkoff
.\"  Extensive changes 2000 by Eric Lee Green <eric@badtux.org>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.
.\"
.TH MTX 1 MTX1.3
.SH NAME
mtx \- control SCSI media changer devices
.SH SYNOPSIS
mtx [-f <scsi-generic-device>] [nobarcode] [invert] [noattach] command [ command ... ]
.SH DESCRIPTION
The
.B mtx
command controls single or multi-drive SCSI media changers such as
tape changers, autoloaders, tape libraries, or optical media jukeboxes.
It can also be used with media changers that use the 'ATTACHED' API,
presuming that they properly report the MChanger bit as required
by the SCSI T-10 SMC specification.
.SH OPTIONS
The first argument, given following
.B -f
, is the SCSI generic device corresponding to your media changer.
Consult your operating system's documentation for more information (for
example, under Linux these are generally /dev/sg0 through /dev/sgXX.
under FreeBSD these are /dev/pass0 through /dev/passX,
under SunOS it may be a file under /dev/rdsk).
.P
The 'invert' option will invert (flip) the media (for optical jukeboxes that
allow such) before inserting it into the drive or returning it to the
storage slot.
.P
The 'noattach' option forces the regular media changer API even if the
media changer incorrectly reported that it uses the 'ATTACHED' API.
.P
The 'nobarcode' option forces the loader to not request barcodes even if
the loader is capable of reporting them.
.P
Following these options there may follow
one or more robotics control
commands. Note that the 'invert' and 'noattach'
options apply to ALL of robotics control
commands.

.SH COMMANDS
.TP 10
.B --version
Report the mtx version number (e.g. mtx 1.2.8) and exit.

.TP 10
.B inquiry
Report the product type (Medium Changer, Tape Drive, etc.), Vendor ID,
Product ID, Revision, and whether this uses the Attached Changer API
(some tape drives use this rather than reporting a Medium Changer on a
separate LUN or SCSI address).
.TP 10
.B noattach
Make further commands use the regular media changer API rather than the
_ATTACHED API, no matter what the "Attached" bit said in the Inquiry info.
Needed with some brain-dead changers that report Attached bit but don't respond
to _ATTACHED API.
.TP 10
.B inventory
Makes the robot arm go and check what elements are in the slots. This
is needed for a few libraries like the Breece Hill ones that do not
automatically check the tape inventory at system startup.
.TP 10
.B status
Reports how many drives and storage elements are contained in the
device. For each drive, reports whether it has media loaded in it, and
if so, from which storage slot the media originated. For each storage
slot, reports whether it is empty or full, and if the media changer
has a bar code, MIC reader, or some other way of uniquely identifying
media without loading it into a drive, this reports the volume tag
and/or alternate volume tag for each piece of media.
For historical reasons drives are numbered from 0 and storage slots are
numbered from 1.
.TP 10
.B load <slotnum> [ <drivenum> ]
Load media from slot <slotnum> into drive <drivenum>. Drive 0 is assumed
if the drive number is omitted.
.TP 10
.B unload [<slotnum>] [ <drivenum> ]
Unloads media from drive <drivenum> into slot <slotnum>. If <drivenum> is
omitted, defaults to drive 0 (as do all commands).
If <slotnum> is omitted, defaults to the slot
that the drive was loaded from. Note that there's currently no way to
say 'unload drive 1's media to the slot it came from', other than to
explicitly use that slot number as the destination.
.TP 10
.B [eepos <operation>] transfer <slotnum> <slotnum>
Transfers media from one slot to another, assuming that your mechanism is
capable of doing so. Usually used to move media to/from an import/export
port. 'eepos' is used to extend/retract the import/export
tray on certain mid-range to high end tape libraries (if, e.g., the tray was
slot 32, you might say say 'eepos 1 transfer 32 32' to extend the tray).
Valid values for eepos <operation>
are 0 (do nothing to the import/export tray), 1, and 2 (what 1 and 2 do varies
depending upon the library, consult your library's SCSI-level
documentation).
.TP 10
.B [eepos <operation>] [invert] [invert2] exchange <slotnum> <slotnum> [<slotnum>]
Move medium from the first slot to the second slot, placing the medium
currently in the second slot either back into the first slot or into the
optional third slot.

.TP 10
.B first  [<drivenum>]
Loads drive <drivenum> from the first slot in the media
changer. Unloads the drive if there is already media in it (note: you
may need to eject the tape using your OS's tape control commands
first).  Note that this command may not be what you want on large
tape libraries -- e.g. on Exabyte 220, the first slot is usually a
cleaning tape. If <drivenum> is omitted, defaults to first drive.

.TP 10
.B last [<drivenum>]
Loads drive <drivenum> from the last slot in the media changer. Unloads
the drive if there is already a tape in it. (Note: you may need to eject
the tape using your OS's tape control commands first).
.TP 10
.B next [<drivenum>]
Unloads the drive and loads the next tape in sequence. If the drive was
empty, loads the first tape into the drive.
.TP 10
.B position <slotnum>
Positions the robot at a specific slot. Needed by some changers to
move to and open the import/export, or mailbox, slot.

.SH AUTHORS
The original 'mtx' program was written by Leonard Zubkoff and extensively
revised for large multi-drive libraries with bar code readers
by Eric Lee Green <eric@badtux.org>. See 'mtx.c' for other contributors.
.SH BUGS AND LIMITATIONS
.P
You may need to do a 'mt offline' on the tape drive to eject the tape
before you can issue the 'mtx unload' command. The Exabyte EZ-17 and 220
in particular will happily sit there snapping the robot arm's claws around
thin air trying to grab a tape that's not there.
.P
For some Linux distributions, you may need to re-compile the kernel to
scan SCSI LUN's in order to detect the media changer. Check
.B lsscsi \-c
to see what's going on.
.P
If you try to unload a tape to its 'source' slot, and said slot is
full, it will instead put the tape into the first empty
slot. Unfortunately the list of empty slots is not updated between
commands on the command line, so if you try to unload another drive to
a full 'source' slot during the same invocation of 'mtx', it will try
to unload to the same (no longer empty) slot and will urp with a SCSI
error.
.P

This program reads the Mode Sense Element Address Assignment Page
(SCSI) and requests data on all available elements. For larger
libraries (more than a couple dozen elements)
this sets a big Allocation_Size in the SCSI command block for the
REQUEST_ELEMENT_STATUS command in order to be able to read the entire
result of a big tape library. Some operating systems may not be able
to handle this. Versions of Linux earlier than 2.2.6, in particular,
may fail this request due to inability to find contiguous pages of
memory for the SCSI transfer (later versions of Linux 'sg' device do
scatter-gather so that this should no longer be a problem).
.P
The
.B eepos
command remains in effect for all further commands on a command
line. Thus you might want to follow
.B eepos 1 transfer 32 32
with
.B eepos 0
as
the next command (which clears the
.B eepos
bits).
.P
Need a better name for 'eepos' command! ('eepos' is the name of the bit
field in the actual low-level SCSI command, and has nothing to do with what
it does).
.P

This program has only been tested on Linux with a limited number of
tape loaders (an IBM TS3100 with bar-code reader and 24 slots,
a dual-drive Exabyte 220 tape library, with bar-code
reader and 21 slots, an Exabyte EZ-17 7-slot autoloader, and a Seagate
DDS-4 autochanger with 6 slots). It may not work on other operating systems
with larger libraries, due to the big SCSI request size.
Please see the project page http://sourceforge.net/projects/mtx for information
on reporting bugs, requesting features and the mailing list for peer support.
.SH HINTS
Under Linux,
.B lsscsi \-g
will tell you what SCSI devices you have and their device names.
.P
Under FreeBSD,
.B camcontrol devlist
will tell you what SCSI devices you
have, along with which
.B pass
device controls them.
.P
Under Solaris, set up your 'sgen' driver so that it'll look for
tape changers (see /kernel/drv/sgen.conf and the sgen man page), type
.B touch /reconfigure
then reboot. You can find your changer in /devices by typing
.B /usr/sbin/devfsadm -C
to clean out no-longer-extant entries in your /devices directory, then
.B find /devices -name \*changer -print
to find the device name. Set the symbolic link
.B /dev/changer
to point
to that device name (if it is not doing so already).
.P
With BRU, set your mount and unmount commands as described on the BRU
web site at http://www.bru.com to move to the next tape when backing up
or restoring. With GNU
.B tar,
see
.B mtx.doc
for an example of how to use
.B tar
and
.B mtx
to make multi-tape backups.

.SH AVAILABILITY
This version of
.B mtx
is currently being maintained by Robert Nelson <robertnelson@users.sourceforge.net> .
The 'mtx' home page is http://mtx.sourceforge.net and the actual code is currently available
there and via SVN from http://sourceforge.net/projects/mtx.
.SH SEE ALSO
.BR mt (1), loaderinfo (1), tapeinfo (1), scsitape (1), scsieject (1)


mtx/CHANGES
CHANGES, mtx 1.3.11
- loaderinfo: Print RequestSense information if an operation fails.
- loaderinfo: Print all information from the Device Configuration Page.
- mtx: Fix problems with a lot of loaders determining the number of elements.
- mtx: Implement Previous command.
- mtx: Fix bugs in Last command.
- scsieject: New tool that handles start/stop, load/unload.
- tapeinfo: Fix inconsistencies in output.
- Fix timeout handling on Linux.
- Merge in fixes from the Debian and FreeBSD distributions.

CHANGES, mtx 1.3.10
- Add support for Microsoft Windows.
- Add support for Sony VGP-XL1B(2) Media Changer.
  Thanks to Will (nodenet at hotmail dot com)
- Add build support for Microsoft Windows using MinGW native and Linux
  cross-compile.
- Add build support for Microsoft Windows using Microsoft Visual Studio 2005.
- Add support for building on x86_64.
- Add more debugging information.
- Eliminate compiler warnings.

CHANGES, mtx 1.3.9
- Cleaned up scsi_linux.c a little to eliminate around 40 lines of code. 
- Change to using SG_IO ioctl rather than write-read interface. This
  should make us a little safer, as well as (on 2.6.x) letting us issue SCSI 
  commands to regular block devices as well as to /dev/sgXX devices. 
  WARNING: Can cause the system to *CRASH* if the SCSI device is in use, due 
  to brokenness inside the Linux kernel! It's always preferable to use the 
  /dev/sgXX device, which has specialness in its buffer handling that bypasses
  some of the brokenness elsewhere in the SCSI subsystem.
- Check SG version before using SG_IO interface, so that if we're run on a
  Linux 2.2 system and we were compiled on Linux 2.4 or above, 
- If a tape is in the drive, make tapeinfo print out its partition info. Seems
  to work on my DAT drive, anyhow (shrug). 
- Minor cleanups in mtxl.c (other cleanups necessary in tapeinfo and etc.). 

CHANGES, mtx 1.3.8
- Forgot about 8 bytes for header
- Handle pedantic loader whose firmware writer spends too much time
  parsing phrases in the SCSI spec and not enough time in the
  real world. 

CHANGES, mtx 1.3.6
- The You Gotta Be F'ing Kidding Release (see rant on mailing list)
- Added AIX support with http://fz.eryx.net/aix/ GSC driver, courtesy
  of Steve Heck. 
- Make __WEIRD_CHAR_SUPPRESS the default to stop barcode garbling
- Fix core dump on invalid inputs in mtx.c
- Add retry with bigger ALLOCATION_LENGTH if 
  BYTE_COUNT_OF_REPORT_AVAILABLE is > than our original estimate. (see rant
  on mailing list). 
- Shut up the low-level SCSI sg_read and sg_write routines, which were poofting
  all over the place on innocuous things like no EAAP. 

CHANGES, mtx 1.3.5
- The I-Need-A-Job Release (see http://badtux.org) :-}
- Change Makefile so that it works w/systems whose 'install' program
  accepts only one argument. That's it. 
- Uncomment the Exabyte hack in the drive parsing code so that it works
  with ATL loaders again. 
- Add hack for ATL stacked loaders to keep them from reporting ghost slots

CHANGES, mtx 1.3.4
- Forward-ported the ATL/Exabyte patches from mtx 1.2
- Added progress indicator to 'tapeinfo' for DAT drives. (PLEASE TEST)
- Purged a few more estinc.com out of the documentation, point to bru.com.

CHANGES, mtx 1.3.3
- Reversed James' long descriptor patch until can figure out what's wrong with
  it.
- Fixed some debug declarations, courtesy of Kevin Wang. 
- Point COMPATIBILITY file at mtx.sourceforge.net.

CHANGES, mtx 1.3.2
- Merged Makefile changes from mtx-1.2 code
- Changes from James Dugal to get all info for newer loaders, fix debugging

CHANGES, mtx 1.3.1
- add an Exchange command

CHANGES, mtx 1.3.0
- New Linux SCSI subsystem interface for 2.4 kernels, ripped boldly 
  from sgtools by Doug Gilbert
- mtx 'next' now skips blank slots, courtesy of Christopher McCrory
- mtx 'unload' now prints an error message telling you that you need to
  eject the tape in the drive prior to unloading it, under certain
  conditions. 
- Started work on 'nsmhack' for controlling NSM jukeboxes. 
- copy_barcode was off by one. 
- 'position to element' command now added, courtesy of Mahlon Stacy

CHANGES, mtx 1.2.15:
- Some Solaris fixes, courtesy of Matt Ward
- Fix URL in .spec file

CHANGES, mtx 1.2.14:
- Fix so it'll work if 0 is result of SCSI open (e.g., in cron jobs on Linux)
- Move changelog to end of .spec file for easier reading
- Added a bit of text to beginning of COMPATIBILITY file

CHANGES, mtx 1.2.13:
- Fixed some autoloader bugs w/autoloaders that don't report an arm. 
- Fixed barcode backoff.
- Added "nobarcode" option
- Increased timeout for 'mtx inventory' to 30 minutes
  (note: may increase this even more if needed, please let me know!)
- Shortened timeout for 'mtx inquiry' to 30 seconds
- tapeinfo now prints SCSI ID/LUN info if available (only on Linux at the
  moment, sigh). 
- update documentation w/new email addresses, updated compile directions,
  various fixes. 

CHANGES, mtx 1.2.12:
- Fix FreeBSD compile bugs
- Fix SGI compile bugs
- Add HP/UX port (I hope!), courtesy of William Smith. 
- Re-wrote ReadElementStatus to make work for %!@# brain dead firmware that
  reports non-existent drives (I hope!). Also has side-effect of now working
  with multiple-arm libraries (though it only sees first arm!).
- Cleaned up all -Wall messages.
- Cleaned up Linux Sparc, installs loaderinfo.1, courtesy of Matt Dainty.
- tapeinfo now reports status of CheckUnitReady.  
- tapeinfo no longer puts out Block Position if CheckUnitReady says 'no'.
- tapeinfo now puts out Density Code and Medium Type/Not Loaded (modification
    of patches sent in by Bob Rahe)

CHANGES, mtx 1.2.11:
- Added a GNU autoconf Makefile.in (still provide a Makefile for your use)
  *WARNING* autoconf is not yet working on all supported OS's! You may need
  to do the old-fashioned 'edit Makefile' bit!
- Changed mtx.h and mtxl.c to include and define various things based upon
  what features autoconf detected (e.g., if 'camlib.h', use FreeBSD-style

mtx/contrib/mtx.py
# Copyright 2000 Enhanced Software Technologies Inc.
# Released under Free Software Foundation's General Public License,
# Version 2 or above
#
# This is an example of how to parse the 'mtx' output from a scripting
# language. 
#
# Routine to call 'mtx' and read status, or
# whatever.
#
# We do this here rather than creating a Python "C" module because:
#  1) Reduces complexity of compile environment
#  2) Easier debugging.
#  3) More in keeping with the Unix philosophy of things.
#  4) Easier to add new functionality. 
#
#

import string
import os
import time # we can do some waiting here...

def readpipe(command):
    result=""
    
    infile=os.popen(command,"r")

    try:
        s=infile.read()
    except:
        s=""
        pass
    if not s:
        return None  # didn't get anything :-(. 
    result=result+s
    return result




# these are returned by the mtx.status routine:
class slotstatus:
    def __init__(self,slotnum,middle,barcode=None):
        middle=string.strip(middle)
        try:
            left,right=string.split(middle," ")
        except:
            left=middle
            right=None
            pass
        # Note: We will not be able to test this at the moment since the
        # 220 has no import/export port!
        if right=="IMPORT/EXPORT":
            self.importexport=1
        else:
            self.importexport=0
            pass
        self.slotnum=int(slotnum) # make sure it's an integer...
        if left=="Full":
            self.full=1
        else:
            self.full=None
            pass
        if not barcode:
            self.voltag=None
        else:
            l=string.split(barcode,"=",1)
            self.voltag=l[1]
            pass
        return
    pass

# Drive status lines have this format:
#Data Transfer Element 0:Full (Storage Element 10 Loaded):VolumeTag = B0000009H
#Data Transfer Element 1:Empty

class drivestatus:
    def __init__(self,slotnum,middle,barcode=None):
        self.slotnum=slotnum
        if middle=="Empty":
            self.full=None
            self.origin=None
            self.voltag=None
            return
        else:
            self.full=1
            pass

        # okay, we know we have a tape in the drive. 
        # split and find out our origin: we will always have
        # an origin, even if the #$@% mtx program had to dig one
        # out of the air:

        l=string.split(middle," ")
        self.origin=int(l[3])

        if not barcode:
            self.voltag=None # barcode of this element.
        else:
            l=string.split(barcode," ")
            self.voltag=string.strip(l[2])
            pass
        return
    pass

# This is the return value for mtx.status. 
class mtxstatus:
    def __init__(self):
        self.numdrives=None
        self.numslots=None
        self.numexport=None
        self.drives=[]  # a list of drivestatus instances.
        self.slots=[]   # a list of slotstatus instances
        self.export=[]  # another list of slotstatus instances
        return
    pass

# call 'mtx' and get barcode info, if possible:
# okay, we now have a string that consists of a number of lines.
# we want to explode this string into its component parts.
# Example format:
# Storage Changer /dev/sgd:2 Drives, 21 Slots
# Data Transfer Element 0:Full (Storage Element '5' Loaded) 
# Data Transfer Element 1:Empty
# Storage Element 1:Full :VolumeTag=CLNA0001
# Storage Element 2:Full :VolumeTag=B0000009
# Storage Element 3:Full :VolumeTag=B0000010
# ....
# What we want to do, then, is:
# 1) Turn it into lines by doing a string.split on newline.
# 2) Split the 1st line on ":" to get left and right.
# 3) Split the right half on space to get #drives "Drives," #slots
# 4) process the drives (Full,Empty, etc.)
# 5) For each of the remaining lines: Split on ':'
# 6) Full/Empty status is in 2)
#
configdir="/opt/brupro/bin"  # sigh. 

def status(device):
    retval=mtxstatus()
    command="%s/mtx -f %s status" % (configdir,device)
    result=readpipe(command)
    if not result:
        return None  # sorry! 
    # now to parse:

    try:
        lines=string.split(result,"\n")
    except:
        return None # sorry, no status!
    
    # print "DEBUG:lines=",lines

    try:
        l=string.split(lines[0],":")
    except:
        return None # sorry, no status!
    
    # print "DEBUG:l=",l
    leftside=l[0]
    rightside=l[1]
    if len(l) > 2:
        barcode=l[3]
    else:
        barcode=None
        pass
    t=string.split(rightside)
    retval.numdrives=int(t[0])
    retval.numslots=int(t[2])
    
    for s in lines[1:]:
        if not s:
            continue # skip blank lines!
        #print "DEBUG:s=",s
        parts=string.split(s,":")
        leftpart=string.split(parts[0])
        rightpart=parts[1]
        try:
            barcode=parts[2]
        except:
            barcode=None
            pass
        #print "DEBUG:leftpart=",leftpart
        if leftpart[0]=="Data" and leftpart[1]=="Transfer":
            retval.drives.append(drivestatus(leftpart[3],rightpart,barcode))
            pass
        if leftpart[0]=="Storage" and leftpart[1]=="Element":
            element=slotstatus(leftpart[2],rightpart,barcode)
            if element.importexport:
                retval.export.append(element)
            else:
                retval.slots.append(element)
                pass
            pass
        continue

    return retval

# Output of a mtx inquiry looks like:
#
#Product Type: Medium Changer
#Vendor ID: 'EXABYTE '
#Product ID: 'Exabyte EZ17    '
#Revision: '1.07'
#Attached Changer: No
#
# We simply return a hash table with these values { left:right } format. 

def mtxinquiry(device):
    command="%s/mtx -f %s inquiry" % (configdir,device)
    str=readpipe(command) # calls the command, returns all its data.

    str=string.strip(str)
    lines=string.split(str,"\n")
    retval={}
    for l in lines:
        # DEBUG #
        l=string.strip(l)
        print "splitting line: '",l,"'"
        idx,val=string.split(l,':',1)
        val=string.strip(val)
        if val[0]=="'":
            val=val[1:-1] # strip off single quotes, sigh. 
            pass 
        retval[idx]=val
        continue
    return retval

# Now for the easy part:

def load(device,slot,drive=0):
    command="%s/mtx -f %s load %s %s >/dev/null " % (configdir,device,slot,drive)
    status=os.system(command)
    return status

def unload(device,slot,drive=0):
    command="%s/mtx -f %s unload %s %s >/dev/null " % (configdir,device,slot,drive)
    return os.system(command)

def inventory(device):
    command="%s/mtx -f %s inventory >/dev/null " % (configdir,device)
    return os.system(command)

def wait_for_inventory(device):
    # loop while we have an error return...
    errcount=0
    while inventory(device):
        if errcount==0:
            print "Waiting for loader '%s'" % device
            pass
        time.sleep(1)
        try:
            s=status(device)
        except:
            s=None
            pass
        if s:
            return 0 # well, whatever we're doing, we're inventoried :-(
        errcount=errcount+1
        if errcount==600:   # we've been waiting for 10 minutes :-(
            return 1 # sorry!
        continue
    return 0  # we succeeded!

# RCS REVISION LOG:
# $Log$
# Revision 1.1  2001/06/05 17:10:51  elgreen
# Initial revision
#
# Revision 1.2  2000/12/22 14:17:19  eric
# mtx 1.2.11pre1
#
# Revision 1.14  2000/11/12 20:35:29  eric
# do string.strip on the voltag
#
# Revision 1.13  2000/11/04 00:33:38  eric
# if we can get an inventory on the loader after we send it 'mtx inventory',
# then obviously we managed to do SOMETHING.
#
# Revision 1.12  2000/10/28 00:04:34  eric
# added wait_for_inventory command
#
# Revision 1.11  2000/10/27 23:27:58  eric
# Added inventory command...
#
# Revision 1.10  2000/10/01 01:06:29  eric
# evening checkin
#
# Revision 1.9  2000/09/29 02:49:29  eric
# evening checkin
#
# Revision 1.8  2000/09/02 01:05:33  eric
# Evening Checkin
#
# Revision 1.7  2000/09/01 00:08:11  eric
# strip lines in mtxinquiry
#
# Revision 1.6  2000/09/01 00:05:33  eric
# debugging
#
# Revision 1.5  2000/08/31 23:46:01  eric
# fix def:
#
# Revision 1.4  2000/08/31 23:44:06  eric
# =->==
#
